Vim Mode 3 - Update hotkey infrastructure to work around various issues with react-hotkeys

tntmarket opened the pull request:
This is part of a larger change to support vim style navigation (#63)

This wraps `react-hotkeys` to make it behave in a more intuitive and correct way.

These are the main use cases that hinder vim mode, and their proposed fix:

| Use Case | Status Quo | Desired Behavior |
| ----------|------------|-------------------|
| Letting go of `shift` before letting go of `V` | React hotkeys thinks the `v` key was never released, blocking all future shortcuts | Pressing `shift`-`v`-`shift up`-`v up` should trigger the `V` hotkey without blocking future hotkeys |
| Binding a hotkey for `J` | Only triggers when _holding_ `shift+j`. If you bind `shift+j`, only the initial _press_ works.  | Should trigger both on the initial press of `shift+j`, and when holding `shift+j`|
| Binding a hotkey for `u` | Native shortcuts that use `u` such as `cmd+u` stop working | Native shortcuts should keep working |
| Binding a hotkey for `Esc` | If another hotkey simulates `Esc` (in order to unfocus a block for example), then it will actually trigger the `Esc` hotkey | If we are simulating a keypress trigger native actions, we probably don't want to trigger our own custom hotkeys |
| Binding a hotkey for a sequence like `g g` | Makes it so plain single hotkeys stop doing anything | Plain single hotkeys should keep working |

I acknowledge that a longer-term solution is better, but figured that this was a good enough band-aid until we figure out what to do about #68

tntmarket committed:
Update hotkey infrastructure to work around various issues with react-hotkeys and provide a more convenient interface that does what you'd expect

tntmarket commented on src/ts/core/common/react-hotkeys-fixed/index.tsx:
```diff
@@ -0,0 +1,91 @@
+     */
+    stopEventPropagationAfterHandling: false,
+    stopEventPropagationAfterIgnoring: false,
+    /**
+     * React Hotkeys treats j/J as different keys, which is buggy when detecting key chords.
+     * For example, if I press these keys:
+     *
+     *     [shift down, j down, shift up, j up]
+     *
+     * React Hotkeys interprets it like so:
+     *
+     *            ---------Shift+J-------
+     *            |                     |
+     *     [Shift+J down, J down, Shift+J up, j up]
+     *                       |                  |
+     *                       ----Not Released----
+     *
+     * React Hotkeys thinks that the "J" key is still down, which breaks all hotkeys until you
+     * Un-focus the browser window (resetting the key history).
+     *
+     * To workaround this issue, we normalize j/J to be the "same key" using the keycode,
+     * That way, a "J" keydown is stored as a "j" keydown in the key history, so a "j" keyup
+     * will "release" it.
+     *
+     * Related issue: https://github.com/greena13/react-hotkeys/issues/249
+     */
+    customKeyCodes: CODE_TO_KEY,
```
This is the issue that `CODE_TO_KEY` from https://github.com/roam-unofficial/roam-toolkit/pull/63 is trying to solve

Stvad commented on src/ts/core/common/react-hotkeys-fixed/allow-holding-shifted-keys.ts:
```diff
@@ -0,0 +1,46 @@
+// A "KeySequence" is a series of key chords, such as "g g" or "control+w j"
```
that's a weird place to have this comment :)

Stvad commented on src/ts/core/common/react-hotkeys-fixed/dont-override-native-hotkeys.tsx:
```diff
@@ -0,0 +1,38 @@
+import {GlobalHotKeys, KeyMap} from 'react-hotkeys'
```
this whole fix thing probably deserves it's own subdirectory in `core` vs being under `common`

Stvad commented on src/ts/core/common/react-hotkeys-fixed/allow-holding-shifted-keys.ts:
```diff
@@ -0,0 +1,46 @@
+ *         The list contains multiple alternatives trigger the same handler.
+ *         See https://github.com/greena13/react-hotkeys#alternative-hotkeys
+ */
+export const allowHoldingShiftedKeys = (keySequence: KeySequence): KeyChord[] | KeySequence => {
```
these functions could use the Specs

Stvad commented on src/ts/core/common/react-hotkeys-fixed/allow-holding-shifted-keys.ts:
```diff
@@ -0,0 +1,46 @@
+const convertUppercaseToLowercasePlusShift = (keyChord: KeyChord): KeyChord => {
+    const keys = keyChord.split('+')
+    const key = keys.pop()
+    const modifiers = keys.filter(key => key.toLowerCase())
```
`filter` ? this seems like it should have been a map maybe?

Stvad commented on src/ts/core/common/react-hotkeys-fixed/allow-holding-shifted-keys.ts:
```diff
@@ -0,0 +1,46 @@
+ */
+const convertUppercaseToLowercasePlusShift = (keyChord: KeyChord): KeyChord => {
+    const keys = keyChord.split('+')
+    const key = keys.pop()
```
I think this should signify more explicitly what it is, maybe `letter` (though it probably does not have to be a letter 🤔)?

tntmarket replied:
I extracted `KeySequence.keySequencesForBothHeldAndInitialPress`. I've been struggling to come up with good names for these hotkey wrappers :/

Stvad commented on src/ts/core/common/react-hotkeys-fixed/dont-override-native-hotkeys.tsx:
```diff
@@ -0,0 +1,38 @@
+ * Don't trigger the single letter shortcuts when pressing native keybindings such as "cmd+u".
+ * See https://github.com/greena13/react-hotkeys/issues/234#issuecomment-612687273
+ */
+const NATIVE_HOTKEYS_TO_IGNORE: KeyChord[] = [
```
this seems too mac specific, and does not seem to cover all the native shortcuts even for mac? also not sure it's very sustainable

Stvad replied:
hmm, just looked at the issue - seems it does not happen for `ctrl` modifier? so this is only `cmd` thing?

Stvad replied:
also this is missing `command+r`

Stvad commented on src/ts/core/common/react-hotkeys-fixed/dont-trigger-hotkeys-with-simulated-presses.ts:
```diff
@@ -0,0 +1,49 @@
+const keyChordMightBeSimulated = (keyChord: KeyChord) =>
+    KEYS_THAT_WE_ALSO_SIMULATE.some(key => keyChord.includes(key))
+
+let executingHandler = 0
```
this feels like it should be a class

Stvad commented on src/ts/core/common/react-hotkeys-fixed/hotkey.ts:
```diff
@@ -0,0 +1,66 @@
+ */
+const clearKeyPressesAfterFinishingKeySequence = (handler: Handler): Handler => async event => {
+    await handler(event)
+    KeyEventManager.getInstance()._clearKeyHistory()
```
is this a global object for all instances of react-hotkeys?

Stvad commented on src/ts/core/common/react-hotkeys-fixed/hotkey.ts:
```diff
@@ -0,0 +1,66 @@
+        return this.keySequence.includes(' ')
+    }
+
+    fixedKeySequence() {
```
Even though we fix them here, I don't think `fixed` should be part of the name. It's the internal detail of implementation

Stvad commented on src/ts/core/common/react-hotkeys-fixed/hotkeys.ts:
```diff
@@ -0,0 +1,46 @@
+    private actionToHotkey: Dictionary<Hotkey>
+
+    static fromKeyMapAndHandlers(actionToKeySequence: Dictionary<KeySequence>, actionToHandler: Dictionary<Handler>) {
+        const actionToHotkey: Dictionary<Hotkey> = {}
+        Object.keys(actionToKeySequence).forEach(action => {
```
nit: consider using `reduce` for this

Stvad commented on src/ts/core/common/react-hotkeys-fixed/hotkeys.ts:
```diff
@@ -0,0 +1,46 @@
+        this.actionToHotkey = actionToHotkey
+    }
+
+    singleChordHotkeys() {
+        return new Hotkeys(
+            pickBy(this.actionToHotkey, hotkey => !hotkey.usesMultipleKeyChords())
+        )
+    }
+
+    multiChordHotkeys() {
+        return new Hotkeys(
+            pickBy(this.actionToHotkey, hotkey => hotkey.usesMultipleKeyChords())
+        )
+    }
```
these don't feel like they conceptually belong inside this object. maybe make them static functions?

tntmarket replied:
I decided to inline these after extracting classes for `KeyChord` / `KeySequence`, since "Hotkey" as the combination of "Key+Handler" didn't seem as useful of a definition anymore.

Stvad reviewed (commented):
the necessity of this PR makes me sad.  
also would benefit from tests

tntmarket committed:
Move react-hotkeys-fixed to core/react-hotkeys

Stvad committed:
Merge pull request #93 from tntmarket/master

Various utilities for browser APIs, generic data manipulation

Stvad committed:
Run tests on pull-requests

Stvad committed:
Merge pull request #94 from tntmarket/master

Fix tests by updating import paths

Stvad committed:
Remove `unsafe-eval` that I accidentally commited before

tntmarket committed:
Extract classes for key chords and key sequences. Add tests

tntmarket commented:
I rebased on master, and added some tests.

I also extracted classes for `KeySequence` and `KeyChord`. I think it's more verbose, but maybe easier to understand than directly manipulating lists and strings.

tntmarket committed:
Use reduce instead of forEach

tntmarket committed:
Don't say `Fixed` in react hotkeys wrapper names

Stvad commented on package.json:
```diff
@@ -18,8 +18,10 @@
     "license": "MIT",
     "dependencies": {
         "@popperjs/core": "^2.4.0",
+        "@types/immutable": "^3.8.7",
```
I believe types should go to dev deps

Stvad commented on tests/ts/common/object.test.ts:
```diff
@@ -0,0 +1,10 @@
+import {zipObjects} from 'src/core/common/object'
+
+describe(zipObjects, () => {
+    it('bundles values matching the same key into a tuple', async () => {
```
add a specification for case when there is mismatched keys?

Stvad commented on src/ts/core/react-hotkeys/key-handler.ts:
```diff
@@ -0,0 +1,55 @@
+ */
+export type Handler = (event: KeyboardEvent) => Promise<any> | undefined
+
+let executingHandler = 0
```
comments didn't seem to get carried over, as file moved 🤔 anyhow - I believe this module should be a class https://github.com/tntmarket/roam-toolkit/pull/2#discussion_r446573189

tntmarket replied:
Are you thinking something like a Command Object style class with an execute method?

```typescript
class KeyHandler {
    private static executingHandler = 0
    private handler: Handler
    private keySequence: KeySequence

    constructor(keySequence: KeySequence, handler: Handler) {
        this.keySequence = keySequence;
        this.handler = handler;
    }

    async execute(event: KeyboardEvent) {
        await this.executeIfTriggeredByUser(event)
        if (this.keySequence.usesMultipleKeyChords()) {
            clearKeyHistory()
        }
    }

    private async executeIfNothingElseIsExecuting(event: KeyboardEvent) {
        if (KeyHandler.executingHandler === 0) {
            await this.executeAndTrack(event)
        }
    }

    private async executeAndTrack(event: KeyboardEvent) {
        KeyHandler.executingHandler += 1
        try {
            await this.handler(event)
        } catch (error) {
            console.error(error)
        }
        KeyHandler.executingHandler -= 1
    }

    /**
     * If we artificially simulate a key press, that keypress should not
     * trigger our own hotkeys.
     *
     * For example, simulating "Esc" to unfocus a block should not trigger
     * our own hotkey for "Esc".
     *
     * @return a decorated version of a handler that does nothing if other
     *         handlers are running
     */
    private async executeIfTriggeredByUser(event: KeyboardEvent) {
        if (this.keySequence.mightBeSimulated()) {
            await this.executeIfNothingElseIsExecuting(event)
        } else {
            await this.executeAndTrack(event)
        }
    }
}
```

Or alternatively, keeping the decorator style, but use a class:

```typescript
class KeyHandler {
    // ...
    private handler: Handler

    constructor(handler: Handler) {
        this.handler = handler;
    }

    preventWhileOtherHandlersAreExecuting(): KeyHandler {
        return new KeyHandler(
            // ... decorate this.handler
        )
    }

    // ... same with other methods
}
```

I personally find decorator+class style less readable, because it adds indirection. I find command object style maybe a bit more readable, but less composable.

Stvad replied:
My first concern was to encapsulate the global state here more clearly.  
I suppose it's kind of already done on a module level. So we kind of have a singleton represented by a module. So maybe my concern is coming from mostly working with languages with different module system..  
But does this approach work well with multiple instances of react-hotkeys? I suppose in our case we want this to be global between 2 instances we create. Not sure it's made clear enough though that it has the global state via the current interface 🤔 

I think I'd probably prefer to stay with the decorator approach as the current handler interface is what we use throughout the project and is what expected by react-hotkeys. So I think even if we use command object - we'd just use it for this specific case and would need to wrap it back to that interface to pass it on to the react-hotkeys

Also unfortunate that we can't make objects invokable.

tntmarket replied:
In order to communicate to readers - _"WARNING, this module uses global state, decorating one handler will affect the behavior of already decorated handlers"_, I'll try the following:

* Include a docstring for `adaptHandlerToReactHotkeys` warning that it uses a shared counter, and recommend that they look at the test cases.
* Inline `adaptHandlerToReactHotkeys`, and rename `dontTriggerWhenKeyPressIsSimulated` => `forbidConcurrentHandling` (naming it based on what it does rather what it's intended purpose is)

Stvad commented on src/ts/core/react-hotkeys/key-sequence.ts:
```diff
@@ -0,0 +1,65 @@
+     * @return a list of alternative key sequences that should all trigger the same handler
+     */
+    private keySequencesForBothHeldAndInitialPress(): KeySequence[] {
+        const keyChord = this.keyChords[0]
+        if (keyChord.isCapitalLetter() && !this.usesMultipleKeyChords()) {
+            return [this, this.map(keyChord => keyChord.convertCapitalToShiftAndLowercase())]
+        }
+
+        return [this.map(keyChord => keyChord.convertCapitalToShiftAndLowercase())]
```
I'm confused on why we need a first special case?

tntmarket replied:
Oh wow, I'm no longer able to reproduce the bug that I thought I needed the first case to solve.

I might've been using different `react-hotkeys` configurations (like `simulateMissingKeyPressEvents` and `ignoreRepeatedEventsWhenKeyHeldDown`) at the time I created this workaround, but I don't need it anymore so I'll just delete it.

So I guess the purpose of this is just reduced to allowing capital letters as shorthand for `shift-*` now.

Stvad replied:
Is it worth it just for that?

tntmarket replied:
I'll cut this functionality out. It'll allow a decent amount of code to be deleted

tntmarket replied:
I decided to keep `KeyChord`, even though it doesn't really have functionality, because I think it helps communicate the taxonomy of `Key*` related objects.

Also I kept `KEY_TO_SHIFTED` because it's still used for shift+clicking hints: https://github.com/tntmarket/roam-toolkit/pull/4

Stvad replied:
> Also I kept KEY_TO_SHIFTED because it's still used for shift+clicking hints: #4

hmm, can't we just add `shift` modifier there?

Stvad commented on src/ts/core/react-hotkeys/key-chord.ts:
```diff
@@ -0,0 +1,52 @@
+        this.modifiers = modifiers
+    }
+
+    isCapitalLetter(): string | undefined {
```
this name seems to imply that it'd return boolean. but it returns the unshifted version of the letter ? so maybe something like `unshiftedKey` property instead?

tntmarket replied:
I'll inline this function, since the other usage is no longer needed due to getting rid of the special case above

Stvad commented on src/ts/core/react-hotkeys/key-chord.ts:
```diff
@@ -0,0 +1,52 @@
+    convertCapitalToShiftAndLowercase(): KeyChord {
+        const unshifted = this.isCapitalLetter()
+        if (unshifted) {
+            return new KeyChord(unshifted, this.modifiers.add('shift'))
```
isn't this supposed to be immutable set?

tntmarket replied:
Yeah, `this.modifiers.add('shift')` returns a copy of the set with `shift` added. I guess it's immutable in the sense that I'm leaving the old reference alone, but I'm still "changing" the set.

I used an immutable set, so`KeyChord` would behave like an immutable value object.

Stvad commented on src/ts/core/react-hotkeys/key-chord.ts:
```diff
@@ -0,0 +1,52 @@
+     * @return the keychord with capitals converted to shift+lowercase, or null if
+     *         the key is already lowercase
+     */
+    convertCapitalToShiftAndLowercase(): KeyChord {
```
this potentially should return a copy with new set of modifiers (will work with immutable set too)

tntmarket replied:
Oh I think that's what I'm doing already?

Stvad replied:
oh darn, indeed 😅

Stvad commented on src/ts/core/react-hotkeys/key-history.ts:
```diff
@@ -0,0 +1,16 @@
+ */
+export const clearKeyPressesAfterFinishingKeySequence = (handler: Handler): Handler => async event => {
+    await handler(event)
+    KeyEventManager.getInstance()._clearKeyHistory()
```
https://github.com/tntmarket/roam-toolkit/pull/2#discussion_r446573863

tntmarket replied:
Upon further investigation, it does seem that `KeyEventManager` is global to all instances of react-hotkeys: https://sourcegraph.com/github.com/greena13/react-hotkeys/-/blob/src/lib/KeyEventManager.js#L23:3

:/

Stvad replied:
that may end up being problematic. add some logging here?

tntmarket replied:
You mean like `console.warn('clearing react-hotkeys history')`?

Stvad replied:
something like that

Stvad commented:
>I also extracted classes for KeySequence and KeyChord  

Nice, I do believe it made things clearer :)

tntmarket committed:
SQUASH hotkeys - move types to devDependencies

tntmarket committed:
SQUASH hotkeys - test cases for other zipObjects cases

tntmarket committed:
SQUASH hotkeys - test cases for other zipObjects cases

tntmarket committed:
SQUASH hotkeys - remove unnecessary case for expanding capital letter into both it's shifted and capital versions

Stvad commented on tests/ts/common/object.test.ts:
```diff
@@ -0,0 +1,22 @@
+        })
+    })
+
+    it('uses undefined when there is no matching key', async () => {
```
hmm, this is kind of unexpected. Not an empty dict?

tntmarket replied:
I'll exclude keys instead of returning tuples with undefined

tntmarket committed:
Clarify what key-handler.ts actually does, and that it uses global state

tntmarket committed:
Delete normalization of capital letters to shift+lowercase, not needed anymore

tntmarket committed:
Log some warnings when clearing react hotkeys history

tntmarket committed:
Exclude keys instead of returning tuples with undefined with zipObjects
